crates/common/src/unit_tests/mod.rs

```diff
@@ -0,0 +1,470 @@
+        let flattened = result.into_flattened_table().unwrap();
+        Ok(flattened.rows[0].clone())
+    }
+
+    async fn run_calculate_entrypoint(
+        &mut self,
+        pre_stack: Vec<U256>,
+    ) -> Result<RainEvalResults, TestRunnerError> {
+        let input_token = pre_stack[0];
+        let output_token = pre_stack[1];
+        let output_cap = pre_stack[2];
+        // let block_number = pre_stack[3];
+
+        let final_bindings = self.get_final_bindings(false);
+
+        let dotrain = Arc::new(self.dotrains.main_dotrain.clone());
+        self.forker
+            .roll_fork(Some(self.test_setup.block_number), None)?;
+        let fork = Arc::new(self.forker.clone());
+        let fork_clone = Arc::clone(&fork);
+        let deployer = Arc::clone(&self.test_setup.deployer);
+        let dotrain = Arc::clone(&dotrain);
+
+        let handle = tokio::spawn(async move {
+            let rainlang_string = RainDocument::compose_text(
+                &dotrain,
+                &["calculate-io"],
+                None,
+                Some(final_bindings),
+            )?;
+
+            let mut context = vec![vec![U256::from(0); 5]; 5];
+            // output cap
+            context[2][0] = output_cap;
+            // input token
+            context[3][0] = input_token;
+            // output token
+            context[4][0] = output_token;
+
+            let args = ForkEvalArgs {
+                rainlang_string,
+                source_index: 0,
+                deployer: deployer.address,
+                namespace: FullyQualifiedNamespace::default(),
+                context,
+                decode_errors: true,
+            };
+            fork_clone
+                .fork_eval(args)
+                .map_err(TestRunnerError::ForkCallError)
+                .await
+        });
+
+        let result: RainEvalResults = vec![handle.await??.into()].into();
+        Ok(result)
+    }
+
+    async fn run_handle_entrypoint(
+        &mut self,
+        calculate_stack: RainEvalResults,
+    ) -> Result<RainEvalResults, TestRunnerError> {
+        let _io_ratio = calculate_stack.results[0].stack[0];
+        let max_output = calculate_stack.results[0].stack[1];
+
+        let final_bindings = self.get_final_bindings(false);
+
+        let dotrain = Arc::new(self.dotrains.main_dotrain.clone());
+        self.forker
+            .roll_fork(Some(self.test_setup.block_number), None)?;
+        let fork = Arc::new(self.forker.clone());
+        let fork_clone = Arc::clone(&fork);
+        let deployer = Arc::clone(&self.test_setup.deployer);
+        let dotrain = Arc::clone(&dotrain);
+
+        let handle = tokio::spawn(async move {
+            let rainlang_string =
+                RainDocument::compose_text(&dotrain, &["handle-io"], None, Some(final_bindings))?;
+
+            let mut context = vec![vec![U256::from(0); 5]; 5];
+
+            // output vault decrease
+            context[4][4] = max_output;
```

Comment:
doesn't it need to be the min of the calculated output and the vault balance set in pre?

Reply:
this is still relevant and a major bug, right?

Reply:
@thedavidmeister you are right, fixing that now